File: skypydb/api/mixins/vector/get_collection.py
"""Mixin that retrieves vector collections for `VectorClient`."""

from typing import Any, Dict, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from skypydb.api.collection import Collection


class GetCollectionMixin:
    """Expose collection retrieval behavior on `VectorClient`."""

    def get_collection(self, name: str) -> "Collection":
        """Return an existing collection by name."""

        collection_info = self._db.get_collection(name)
        if collection_info is None:
            raise ValueError(f"Collection '{name}' not found")

        if name in self._collections:
            return self._collections[name]

        from skypydb.api.collection import Collection

        collection = Collection(
            db=self._db,
            name=name,
            metadata=collection_info.get("metadata"),
        )
        self._collections[name] = collection
        return collection

    def get_or_create_collection(
        self,
        name: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> "Collection":
        """Return a collection, creating it when it does not exist."""

        collection_info = self._db.get_or_create_collection(name, metadata)

        if name in self._collections:
            return self._collections[name]

        from skypydb.api.collection import Collection

        collection = Collection(
            db=self._db,
            name=name,
            metadata=collection_info.get("metadata"),
        )
        self._collections[name] = collection
        return collection

File: skypydb-js/src/api/operations/vector/get_collection.ts
import type { Metadata } from "../../../types";
import { Collection } from "../../collection";
import { CreateCollectionMixin } from "./create_collection";
import type { VectorDatabase } from "../../../database/vector_database";

export class GetCollectionMixin extends CreateCollectionMixin {
  async get_collection(name: string): Promise<Collection> {
    const collection_info = this._db.get_collection(name);
    if (!collection_info) {
      throw new Error(`Collection '${name}' not found`);
    }

    const cached = this._collections.get(name);
    if (cached) {
      return cached;
    }

    const collection = new Collection(this._db, name, collection_info.metadata);
    this._collections.set(name, collection);
    return collection;
  }

  async get_or_create_collection(
    name: string,
    metadata?: Metadata,
  ): Promise<Collection> {
    const collection_info = this._db.get_or_create_collection(name, metadata);
    const cached = this._collections.get(name);
    if (cached) {
      return cached;
    }

    const collection = new Collection(this._db, name, collection_info.metadata);
    this._collections.set(name, collection);
    return collection;
  }
}
